openvdb_ax/openvdb_ax/codegen/StandardFunctions.cc

```diff
@@ -0,0 +1,1982 @@
+
+        assert(m1.size() == 9 || m1.size() == 16);
+        assert(ptrs.size() == m1.size());
+        assert(ptrs.size() == m2.size());
+        const size_t dim = m1.size() == 9 ? 3 : 4;
+
+        llvm::Value* e3 = nullptr, *e4 = nullptr;
+        for (size_t i = 0; i < dim; ++i) {
+            const size_t row = i*dim;
+            for (size_t j = 0; j < dim; ++j) {
+                llvm::Value* e1 = binaryOperator(m1[0+row], m2[j], ast::tokens::MULTIPLY, B);
+                llvm::Value* e2 = binaryOperator(m1[1+row], m2[dim+j], ast::tokens::MULTIPLY, B);
+                if (dim >=3) e3 = binaryOperator(m1[2+row], m2[(dim*2)+j], ast::tokens::MULTIPLY, B);
+                if (dim >=4) e4 = binaryOperator(m1[3+row], m2[(dim*3)+j], ast::tokens::MULTIPLY, B);
+                e1 = binaryOperator(e1, e2, ast::tokens::PLUS, B);
+                if (dim >=3) e1 = binaryOperator(e1, e3, ast::tokens::PLUS, B);
+                if (dim >=4) e1 = binaryOperator(e1, e4, ast::tokens::PLUS, B);
+                B.CreateStore(e1, ptrs[row+j]);
+            }
+        }
+
+        return nullptr;
+    };
+
+    static auto mmmult = [](auto out, auto mat2, auto mat1) {
+        *out = (*mat1) * (*mat2);
+    };
+
+    using MMMultM3D = void(openvdb::math::Mat3<double>*, openvdb::math::Mat3<double>*, openvdb::math::Mat3<double>*);
+    using MMMultM3F = void(openvdb::math::Mat3<float>*, openvdb::math::Mat3<float>*, openvdb::math::Mat3<float>*);
+    using MMMultM4D = void(openvdb::math::Mat4<double>*, openvdb::math::Mat4<double>*, openvdb::math::Mat4<double>*);
+    using MMMultM4F = void(openvdb::math::Mat4<float>*, openvdb::math::Mat4<float>*, openvdb::math::Mat4<float>*);
+
+    return FunctionBuilder("mmmult")
+        .addSignature<MMMultM3D, true>(generate, (MMMultM3D*)(mmmult))
+        .addSignature<MMMultM3F, true>(generate, (MMMultM3F*)(mmmult))
+        .addSignature<MMMultM4D, true>(generate, (MMMultM4D*)(mmmult))
+        .addSignature<MMMultM4F, true>(generate, (MMMultM4F*)(mmmult))
+        .setArgumentNames({"a", "b"})
+        .addParameterAttribute(0, llvm::Attribute::NoAlias)
+        .addParameterAttribute(0, llvm::Attribute::WriteOnly)
+        .addParameterAttribute(1, llvm::Attribute::ReadOnly)
+        .addParameterAttribute(2, llvm::Attribute::ReadOnly)
+        .addFunctionAttribute(llvm::Attribute::NoUnwind)
+        .addFunctionAttribute(llvm::Attribute::InlineHint)
+        .setConstantFold(op.mConstantFoldCBindings)
+        .setPreferredImpl(op.mPrioritiseIR ? FunctionBuilder::IR : FunctionBuilder::C)
+        .setDocumentation("Multiplies two matrices together and returns the result")
+        .get();
+}
+
+inline FunctionGroup::Ptr axpolardecompose(const FunctionOptions& op)
+{
+    static auto polardecompose = [](auto in, auto orth, auto symm) -> bool {
+        bool success = false;
+        try {
+            success = openvdb::math::polarDecomposition(*in, *orth, *symm);
```

Comment:
Does this do negative determinant checking?

Reply:
Added more documentation